client/src/components/Draftail/CommentableEditor/CommentableEditor.js

```diff
@@ -20,21 +20,60 @@ function usePrevious(value) {
 class DraftailInlineAnnotation {
     constructor(field) {
       this.field = field;
-      this.inlineRefs = new Set()
+      this.inlineRefs = new Map();
+      this.focusedBlockKey = '';
+      this.cachedMedianRef = null;
     }
-    addRef(ref) {
-      this.inlineRefs.add(ref);
+    addRef(ref, blockKey) {
+      this.inlineRefs.set(ref, blockKey);
+
+      // We're adding a ref, so remove the cached median refs - this needs to be recalculated
+      this.cachedMedianRef = null;
     }
     removeRef(ref) {
       this.inlineRefs.delete(ref);
+
+      // We're deleting a ref, so remove the cached median refs - this needs to be recalculated
+      this.cachedMedianRef = null;
+    }
+    setFocusedBlockKey(blockKey) {
+      this.focusedBlockKey = blockKey;
     }
-    getDesiredPosition() {
-      const nodeTops = Array.from(this.inlineRefs).map(ref => ref.current).filter(node => node).map(node => node.getBoundingClientRect().top)
-      if (nodeTops.length > 0) {
-        return nodeTops.reduce((a,b) => a + b, 0)/nodeTops.length + document.documentElement.scrollTop
+    static getHeightForRef(ref) {
+      return ref.current.getBoundingClientRect().top
+    }
+    static getMedianRef(refArray) {
+      const refs = refArray.sort((a, b) => this.getHeightForRef(a) - this.getHeightForRef(b))
+      const length = refs.length
+      if (length > 0) {
+        return refs[Math.ceil(length/2 - 1)]
       }
+      return null
+    }
+    getDesiredPosition(focused = false) {
+      // The comment should always aim to float by an annotation, rather than between them, so uses the median height decorator ref
```

Comment:
"median height decorator ref" doesn't make a lot of sense to me!

Does this sound accurate?

> The comment should always aim to float by an annotation, rather than between them, so calculate which annotation is the median one by height and float the comment by that

Reply:
Changed, thanks!